File: prototype/rails/src/test/domain/escrow/ledger_movement_test.rb
require "test_helper"

module Domain
  module Escrow
    class LedgerMovementTest < ActiveSupport::TestCase
      test "a hold adds the net to escrow" do
        movement = LedgerMovement.hold(Money.from_cents(40_500))

        assert_equal LedgerEntryType::HELD, movement.entry_type
        assert_equal 40_500, movement.amount.cents
      end

      test "a release adds the net to what is available" do
        movement = LedgerMovement.release(Money.from_cents(40_500))

        assert_equal LedgerEntryType::RELEASED, movement.entry_type
        assert_equal 40_500, movement.amount.cents
      end

      test "a payout leaves the ledger so it is negative" do
        movement = LedgerMovement.payout(Money.from_cents(40_500))

        assert_equal LedgerEntryType::PAID_OUT, movement.entry_type
        assert_equal(-40_500, movement.amount.cents)
      end
    end
  end
end

File: prototype/node/src/app/core/escrow/ledger-movement.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { holdMovement, payoutMovement, refundMovement, releaseMovement } from './ledger-movement.ts'
import { cents } from '../money.ts'

test('a hold adds the net to escrow', () => {
  const movement = holdMovement(cents(40_500))

  assert.equal(movement.entryType, 'held')
  assert.equal(movement.amountCents, 40_500)
})

test('a release adds the net to what is available', () => {
  const movement = releaseMovement(cents(40_500))

  assert.equal(movement.entryType, 'released')
  assert.equal(movement.amountCents, 40_500)
})

test('a payout leaves the ledger entry negative', () => {
  const movement = payoutMovement(cents(40_500))

  assert.equal(movement.entryType, 'paid_out')
  assert.equal(movement.amountCents, -40_500)
})

test('a payout of zero stays a positive zero', () => {
  const movement = payoutMovement(cents(0))

  assert.equal(Object.is(movement.amountCents, -0), false)
})

test('a refund leaves the ledger entry negative', () => {
  const movement = refundMovement(cents(40_500))

  assert.equal(movement.entryType, 'refunded')
  assert.equal(movement.amountCents, -40_500)
})

test('a refund of zero stays a positive zero', () => {
  const movement = refundMovement(cents(0))

  assert.equal(Object.is(movement.amountCents, -0), false)
})
